File: backend/app/schemas/agent.py
"""租房推荐 Agent —— 请求/响应 schema"""
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.property import PropertySearchResult


# ── 会话 ──────────────────────────────────────────────────────────

class AgentSessionResponse(BaseModel):
    session_id: int
    session_uuid: str
    cart_id: int
    title: str | None = None


class AgentSessionSummary(BaseModel):
    """Agent 对话记录列表项。"""

    session_id: int
    session_uuid: str
    title: str | None = None
    status: str
    message_count: int = 0
    last_message: str | None = None
    created_at: datetime
    updated_at: datetime


class AgentSessionListResponse(BaseModel):
    items: list[AgentSessionSummary] = Field(default_factory=list)
    total: int = 0


class AgentHistoryMessage(BaseModel):
    """可回放的一条 Agent 历史消息。"""

    id: int
    session_id: int
    role: str
    content: str
    metadata: dict | None = None
    created_at: datetime


class AgentHistoryResponse(BaseModel):
    items: list[AgentHistoryMessage] = Field(default_factory=list)
    has_more: bool = False


# ── 消息 ──────────────────────────────────────────────────────────

class AgentFilters(BaseModel):
    """结构化筛选条件。

    前端 filter bar 提供前 6 个基础字段；LLM 从自然语言中提取
    amenities / room_type / poi_requirements 等硬约束字段。
    """
    # ── 基础字段（前端 filter bar） ──
    country: str | None = None
    currency: str | None = None
    district: str | None = None
    price_min: float | None = Field(default=None, ge=0)
    price_max: float | None = Field(default=None, ge=0)
    bedrooms: int | None = Field(default=None, ge=0)
    property_type: str | None = None

    # ── 硬约束字段（LLM 从 NL 提取 + 前端可选） ──
    amenities: list[str] | None = None          # 设施硬要求，如 ["宠物友好", "独立厨房"]
    room_type: str | None = None                # 房型：studio/ensuite/1bed/2bed/3bed+/shared
    bathrooms: int | None = Field(default=None, ge=0)  # 卫生间数
    area_min: float | None = Field(default=None, ge=0)  # 最小面积
    area_max: float | None = Field(default=None, ge=0)  # 最大面积
    min_lease_months: int | None = Field(default=None, ge=1)  # 最短租期
    max_lease_months: int | None = Field(default=None, ge=1)  # 最长租期
    available_from: str | None = None           # 可入住时间（YYYYMM 格式）

    # ── 周边配套硬约束（LLM 提取） ──
    poi_requirements: list[dict] | None = None  # [{"type": "地铁站", "max_distance_m": 500}, ...]

    # ── 通勤硬约束（LLM 提取） ──
    commute_mode: str | None = None             # walking/bicycling/driving/transit
    commute_minutes: int | None = None          # 通勤时间上限（分钟）

    # ── 机构/学校 ──
    institution: str | None = None              # 大学/机构名
    female_only: bool | None = None              # 是否仅限女生

    # ── 元数据（LLM 标注哪些是硬约束、哪些是软偏好） ──
    hard_filters: list[str] | None = None       # 标记为硬约束的字段名，如 ["amenities", "room_type"]
    soft_preferences: list[str] | None = None   # 标记为软偏好的字段名，如 ["price", "district"]


class AgentMessageRequest(BaseModel):
    """Agent 消息请求（extra="ignore" 保证旧前端发 mode 等字段不报错）"""
    # 长需求清单在查询理解与候选打包阶段会被压缩；此处只保留合理的防滥用上限。
    message: str = Field(min_length=1, max_length=20_000)
    filters: AgentFilters | None = None
    # 搜索页当前条件作为对话上下文；自然语言新条件可覆盖它。
    context_filters: AgentFilters | None = None
    compare_property_ids: list[int] | None = None  # 前端候选清单勾选后传，触发对比意图
    # mode 仅保留旧页面兼容；Dispatcher 会把未知值归一化为 auto。
    mode: str | None = "auto"

    model_config = ConfigDict(extra="ignore")


class AgentMemoryResponse(BaseModel):
    """用户显式保存、可跨会话复用的长期偏好。"""

    preferences: AgentFilters = Field(default_factory=AgentFilters)
    updated_at: datetime | None = None


class AgentMemoryUpdateRequest(BaseModel):
    """显式保存长期偏好；replace=false 时只覆盖传入字段。"""

    preferences: AgentFilters
    replace: bool = False


class AgentRecommendation(BaseModel):
    property_id: int
    rank: int = 0
    final_score: float = 0.0
    score_breakdown: dict[str, float] = Field(default_factory=dict)
    match_reason: str = ""
    pros: list[str] = Field(default_factory=list)
    cons: list[str] = Field(default_factory=list)
    property: PropertySearchResult
    poi_distances: dict[str, int] | None = None   # 周边设施距离，如 {"地铁":350, "超市":200}
    source_metadata: dict = Field(default_factory=dict)  # 字段来源与缺失标记


class AgentLink(BaseModel):
    """回复中附带的站内页面深链"""
    label: str
    to: str


class ThinkingStep(BaseModel):
    """专家模式 Agent 执行步骤"""
    agent_id: str
    agent_name: str
    status: str  # "pending" | "running" | "success" | "error"
    summary: str = ""  # 简短摘要
    duration_ms: int = 0


class GuidedOption(BaseModel):
    """渐进选房引导 chip"""
    label: str = ""
    message: str = ""                               # 点击后作为消息发送
    filter_patch: dict | None = None                # 点击后合并到 filters 的结构化 patch
    kind: str = ""                                  # poi / budget / amenity
    icon: str = ""


class AgentSource(BaseModel):
    """用户可见的数据来源状态。"""
    label: str
    status: str = "verified"


class QueryRewriteInfo(BaseModel):
    """查询改写摘要；不包含模型思维过程。"""
    original: str
    rewritten: str
    kind: str = "exact"
    used_llm: bool = False


class ReferenceResolutionInfo(BaseModel):
    """“第几套/最便宜”等指代的解析结果。"""
    resolved_ids: list[int] = Field(default_factory=list)
    labels: list[str] = Field(default_factory=list)
    unresolved: list[str] = Field(default_factory=list)


class AgentStateSummary(BaseModel):
    """当前对话状态，供前端显示需求 chips。"""
    stage: str = "explore"
    filters: dict = Field(default_factory=dict)
    chips: list[dict[str, str]] = Field(default_factory=list)


class AgentMessageResponse(BaseModel):
    reply: str
    intent: str
    recommendations: list[AgentRecommendation] = Field(default_factory=list)    # 全部匹配房源（"查看所有"展开用）
    top_picks: list[AgentRecommendation] = Field(default_factory=list)          # 精选 Top 3（首屏卡片）
    cart_changed: bool = False
    ai_available: bool = True
    quick_replies: list[str] = Field(default_factory=list)   # 后续建议 chips（点击即作为消息发送）
    links: list[AgentLink] = Field(default_factory=list)     # 站内页面深链按钮
    thinking_steps: list[ThinkingStep] = Field(default_factory=list)  # 专家模式 Agent 执行步骤
    guided_options: list[GuidedOption] = Field(default_factory=list)  # 渐进选房引导 chips
    raw_intent: str = "general"              # 后端内部统一意图
    stage: str = "explore"
    sources: list[AgentSource] = Field(default_factory=list)
    relaxation_trace: list[dict] = Field(default_factory=list)
    query_rewrite: QueryRewriteInfo | None = None
    reference_resolution: ReferenceResolutionInfo | None = None
    state_summary: AgentStateSummary | None = None
    # 只包含本轮适合回填普通搜索栏的明确条件；None 表示清除对应条件。
    filter_patch: dict = Field(default_factory=dict)


class FaqChip(BaseModel):
    """FAQ 快捷入口 chip"""
    id: str
    chip: str


# ── 购物车 ────────────────────────────────────────────────────────

class CartItemAddRequest(BaseModel):
    property_id: int
    reason: str | None = None


class CartItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    property_id: int
    reason: str | None = None
    created_at: datetime
    property: PropertySearchResult


class CartRead(BaseModel):
    id: int
    session_id: int | None = None
    items: list[CartItemRead] = []


# ── 对比 ──────────────────────────────────────────────────────────

class CompareRequest(BaseModel):
    """对比请求。property_ids 为空/缺省时对比整个购物车。

    priority 决定加权评分的权重：balanced 均衡 / budget 预算优先 /
    commute 通勤优先 / space 空间优先（非法值按 balanced 处理）。
    """
    property_ids: list[int] | None = None
    priority: str | None = None


class CompareItem(BaseModel):
    property_id: int
    title: str = ""
    pros: list[str] = []
    cons: list[str] = []
    score: int = 0                                  # 系统确定性加权得分（非 LLM 打分）
    score_breakdown: dict[str, int] | None = None   # 分项：price/commute/space/rating
    best_for: str = ""
    commute: str | None = None                      # 如 "最近交通站点约500m"
    rating: float | None = None                     # 机构真实评价均分（1-5）
    review_count: int = 0
    property: PropertySearchResult | None = None


class CompareResponse(BaseModel):
    summary: str
    items: list[CompareItem]
    recommendation: str
    ai_available: bool = True
    priority: str = "balanced"

File: frontend/src/types/agent.ts
// Matches backend: app/schemas/agent.py
import type { PropertySearchResult, PropertyType } from '@/types/property'

export interface AgentSession {
  session_id: number
  session_uuid: string
  cart_id: number
  title: string | null
}

/** Agent 对话记录列表项。 */
export interface AgentSessionSummary {
  session_id: number
  session_uuid: string
  title: string | null
  status: string
  message_count: number
  last_message: string | null
  created_at: string
  updated_at: string
}

export interface AgentSessionListResponse {
  items: AgentSessionSummary[]
  total: number
}

/** 从历史会话回放的消息；结构化结果保存在 metadata 中。 */
export interface AgentHistoryMessage {
  id: number
  session_id: number
  role: 'user' | 'assistant'
  content: string
  metadata: Record<string, unknown> | null
  created_at: string
}

export interface AgentHistoryResponse {
  items: AgentHistoryMessage[]
  has_more: boolean
}

/** 跨会话长期记忆。 */
export interface AgentMemory {
  preferences: AgentFilters
  updated_at?: string | null
}

export interface AgentFilters {
  country?: string | null
  currency?: string | null
  district?: string | null
  price_min?: number | null
  price_max?: number | null
  bedrooms?: number | null
  property_type?: PropertyType | null
  /** 设施硬要求（引导 chip「要独立卫浴」注入） */
  amenities?: string[] | null
  bathrooms?: number | null
  room_type?: string | null
  institution?: string | null
  commute_mode?: string | null
  commute_minutes?: number | null
  area_min?: number | null
  area_max?: number | null
  min_lease_months?: number | null
  max_lease_months?: number | null
  available_from?: string | null
  female_only?: boolean | null
  /** 渐进选房累积的周边偏好：[{type: "transit"}, ...] */
  poi_requirements?: { type: string; max_distance_m?: number }[] | null
}

/** 渐进选房引导 chip：点击后 filter_patch 并入累积 filters 再重发 */
export interface GuidedOption {
  label: string
  message: string
  filter_patch?: Record<string, unknown> | null
  kind: string
  icon: string
}

export interface AgentMessageRequest {
  message: string
  filters?: AgentFilters | null
  /** 普通搜索页当前条件；自然语言本轮修改可以覆盖这些上下文。 */
  context_filters?: AgentFilters | null
  compare_property_ids?: number[]  // 候选清单勾选后传，触发对比意图
  /** 兼容旧页面模式值；当前智能主链路由后端统一归一化。 */
  mode?: string | null
}

export type AgentIntent =
  | 'recommend'
  | 'search'
  | 'add_to_cart'
  | 'remove_from_cart'
  | 'manage_cart'
  | 'compare'
  | 'compare_cart'
  | 'faq'
  | 'general'

export interface AgentRecommendation {
  property_id: number
  rank: number
  final_score: number
  score_breakdown: Record<string, number>
  match_reason: string
  pros: string[]
  cons: string[]
  property: PropertySearchResult
  /** 周边设施最近距离（米）：{"transit":350, "supermarket":200} */
  poi_distances?: Record<string, number> | null
  /** 后端事实来源与缺失字段，Grounded Answer 使用 */
  source_metadata: Record<string, unknown>
}

export interface AgentSource {
  label: string
  status: 'verified' | 'missing' | string
}

export interface QueryRewriteInfo {
  original: string
  rewritten: string
  kind: 'exact' | 'relative' | 'reference' | 'exploratory' | string
  used_llm: boolean
}

export interface AgentStateChip {
  key: string
  label: string
}

export interface AgentStateSummary {
  stage: string
  filters: Record<string, unknown>
  chips: AgentStateChip[]
}

export interface ReferenceResolutionInfo {
  resolved_ids: number[]
  labels: string[]
  unresolved: string[]
}

/** 回复中附带的站内页面深链 */
export interface AgentLink {
  label: string
  to: string
}

/** FAQ 快捷入口 chip */
export interface FaqChip {
  id: string
  chip: string
}

/** 可验证的处理动作摘要；不包含模型思维链。 */
export interface ThinkingStep {
  agent_id: string
  agent_name: string
  status: 'pending' | 'running' | 'success' | 'error'
  summary: string
  duration_ms: number
}

export interface AgentMessageResponse {
  reply: string
  intent: AgentIntent
  recommendations: AgentRecommendation[]   // 全部匹配房源（"查看所有"展开使用）
  top_picks: AgentRecommendation[]          // 精选 Top 3（首屏卡片）
  cart_changed: boolean
  ai_available: boolean
  quick_replies: string[]
  links: AgentLink[]
  thinking_steps: ThinkingStep[]
  /** 渐进选房引导 chips（点击携带 filter_patch 收窄） */
  guided_options: GuidedOption[]
  raw_intent: string
  stage: string
  sources: AgentSource[]
  relaxation_trace: Record<string, unknown>[]
  query_rewrite: QueryRewriteInfo | null
  reference_resolution: ReferenceResolutionInfo | null
  state_summary: AgentStateSummary | null
  /** 本轮可安全同步到普通搜索栏的明确条件；null 表示清除。 */
  filter_patch: Record<string, unknown>
}

export interface AgentStreamMeta extends Partial<AgentMessageResponse> {
  event?: 'status' | 'result'
}

export interface CartItem {
  id: number
  property_id: number
  reason: string | null
  created_at: string
  property: PropertySearchResult
}

export interface Cart {
  id: number
  session_id: number | null
  items: CartItem[]
}

/** 对比优先级：决定加权评分的权重 */
export type ComparePriority = 'balanced' | 'budget' | 'commute' | 'space'

export interface CompareItem {
  property_id: number
  title: string
  pros: string[]
  cons: string[]
  /** 系统确定性加权得分（非 LLM 打分，可复现） */
  score: number
  /** 分项得分：price/commute/space/rating */
  score_breakdown: Record<string, number> | null
  best_for: string
  /** 如 "最近交通站点约500m"（来自 POI 数据） */
  commute: string | null
  /** 机构真实评价均分（1-5），无评价为 null */
  rating: number | null
  review_count: number
  property: PropertySearchResult | null
}

export interface CompareResponse {
  summary: string
  items: CompareItem[]
  recommendation: string
  ai_available: boolean
  priority: ComparePriority
}

/** 聊天气泡（前端本地状态） */
export interface AgentChatMessage {
  /** 后端消息 ID；本地欢迎语和发送中的消息没有 ID。 */
  id?: number
  role: 'user' | 'assistant'
  content: string
  /** 精选 Top 3 房源（首屏横向卡片） */
  topPicks?: AgentRecommendation[]
  /** 全部匹配房源（"查看所有"按钮跳转搜索页使用） */
  allRecommendations?: AgentRecommendation[]
  /** 该条 AI 消息附带的推荐房源，内联渲染成横条 */
  recommendations?: AgentRecommendation[]
  /** 该轮是否有 AI 分析（用于横条上的降级提示） */
  aiAvailable?: boolean
  /** 后续建议 chips（点击即作为消息发送） */
  quickReplies?: string[]
  /** 站内页面深链按钮 */
  links?: AgentLink[]
  /** 专家模式思考步骤 */
  thinkingSteps?: ThinkingStep[]
  /** 渐进收窄/零结果放宽建议 */
  guidedOptions?: GuidedOption[]
  /** 当前会话已记住的条件 */
  stateSummary?: AgentStateSummary | null
  /** 本轮查询改写摘要 */
  queryRewrite?: QueryRewriteInfo | null
  /** 回答使用的数据来源 */
  sources?: AgentSource[]
  /** 该轮已同步或可同步到普通搜索栏的条件。 */
  filterPatch?: Record<string, unknown>
  /** 是否仍在接收流式文本 */
  streaming?: boolean
  /** 固定欢迎语标识，历史回放时不会重复写入后端。 */
  isWelcome?: boolean
}
